### DISCUSSION TITLE
steady state controller in Campbell Diagram

### DISCUSSION THREAD

**Discussion by tchatte3** [OpenFAST/openfast#876] [Q&A]
Hi @ebranlard 

So far I was using NoServo option for Campbell analysis with op specifications coming from [wind speed, Blade pitch, Rotor rpm]. Now, I am starting to implement the Servo option with OP's = [wind speed, Blade pitch, rotor rpm, generator-torque].

I see in Campbell source codes the generator-torque at the OP points were implemented through the matlab-wrapper through the Servodyn rated generator torque in VSCntrl = 1. In order to make sure that the generator torque was not going to motoring conditions (-ve), I also implemented the rated generator speed, and VS_Rgn2k variables at the 21/2 region as follows.

if opts.CompServo > 0 
        paramSD_mod = SetFASTPar(paramSD,'VS_RtTq',OP.GeneratorTorque(iOP));
%============= Added by Tanmoy Chatterjee, 07/07/2020 =================%        
        VS_Rgn2K = OP.GeneratorTorque(iOP)/(1.005*OP.RotorSpeed(iOP)^2); % Added by Tanmoy Chatterjee, 07/07/2020
        paramSD_mod = SetFASTPar(paramSD_mod,'VS_Rgn2K',VS_Rgn2K); % Added by Tanmoy Chatterjee, 07/07/2020
        paramSD_mod = SetFASTPar(paramSD_mod,'VS_RtGnSp',OP.RotorSpeed(iOP));
 %=======================================================================%  
    
While I was able to run the simulations without the generator going to Motoring, this seems to be a fundamental issue, as the OP points are deviating by10% of what I would expect them to be. 

Do you have suggestions/guidance as to how I should approach steady-state controller design.

Best Regards,

**Comment by ebranlard** [comment]
Hi @tchatte3 

I'm glad your hacking your way into the code. To be honest I haven't done much linearization with controller myself. But here are some suggestions:

- When linearizing with a controller, the "trim" option will be extremely helpful to get your desired operating points. For this you need the "dev" version of openfast, and add some lines in the "fst" file (but I think you have already tried that)
- The `writeLinearizationFiles` script will try to use the Trim option if you have these new input file. The rated torque is estimated as the max generator torque of your operating points. If you are below rated torque, it will use `TrimCase=2`, trying to adjust the generator torque to have the rotor speed you desire. Above rated, it will use `TrimCase=3`, adjusting the pitch to get the operating point rotor speed .  A naive logic is used there, and you might want to change the default values of the `TrimGain` optional variables
- The values you changed `VS_Rgn2K` and `VS_RtGnSp` are usually constants. So instead of changing them in the scripts, I would advise to set them in your "template" servodyn file. 
- I think you need a pitch controller to be able to do the linearization above rated, that might be your main issue. You can try to adapt the DISCON controller from the NREL5MW, or use the ROSCO controller. https://github.com/NREL/ROSCO

I hope that helps,

Emmanuel

**Comment by tchatte3** [comment]
@ebranlard I see your point. So, it might be that instead of relying on the steady state controller I can use Discon type of controller for openfast. I agree, getting the blade pitch correct in the OP points list is critical.

I really appreciate your feedback here to start in the right direction. 

Best Regards,
